webpack/scenes/RedHatRepositories/components/RepositorySetRepository.js

```diff
@@ -63,13 +65,30 @@ class RepositorySetRepository extends Component {
     };
   }
 
+  yStreamHelp = () => (
+    <span translate>
+      This repository is not suggested. Please see additional&nbsp;
+      <a href="https://access.redhat.com/articles/1586183">documentation</a>
+      &nbsp;prior to use.
+    </span>
+  );
+
+  shouldDeemphasize = ({type, releasever}) => type !== 'kickstart' && yStream(releasever);
+
+  repositoryHeading = (props = {arch, releasever}) => (
+    <span>
+      {props.arch} {props.releasever}
+      {this.shouldDeemphasize(props) ? (<FieldLevelHelp content={this.yStreamHelp()} />) : null}
+    </span>
+  );
```

Comment:
This feels more difficult to read now IMO.

Reply:
ya, I would agree, If you want to keep the functions that return JSX (`yStreamHelp` and `repositoryHeading`) in the `render()` function, that is fine. React doesn't give a lot of guidance in this area (afaict) so its really up to our preference and feel.

I didn't mean to drag things out with my suggestions, I'm just trying to avoid crowding the `render()` function [like we do elsewhere](https://git.io/fAckL)

Reply:
I see what you mean. Honestly, I'd really like to see this not made into a function and live in render itself but I see how that would not be very clean based on what ListView.Item is expecting, but I can live with this way :)

Reply:
I think a good rule of thumb is, if you have functions that need props, let's put them in render, else, put them on the component itself. I honestly don't see why defining functions in render is a bad idea (convince me!). It's easy to read, and the limited scope makes it easy to see that they're for the render function and nothing else.

Reply:
My thoughts are:

- How do we test those functions inside of render?

- Shoving everything needing props into render feels like we are going back to writing jQuery where there is no real separation of concerns

- How much responsibility and behavior is appropriate for the render method? Seems like it should be concerned primarily with rendering things (per the name) and not how to render them (hiding functions inside). I would say this is a violation of the single responsibility principle.

Reply:
+1 to all of @jturel's points and I would add that `render()` is called on each render of the page, so you are redefining functions on a render.

I mentioned keeping the functions that return JSX inside render as they are part of what is shown on screen, so technically (IMO) could fall under the responsibility of render. I'm also fine with them being declared outside of render as well like you have.

I wouldn't worry too much about props being passed around, thats pretty common. Is there a reason to not use this syntax to deconstruct props into variables?
```js
  repositoryHeading = ({arch, releasever}) => (
    <span>
      {arch} {releasever}
```

Reply:
> you are redefining functions on a render

Yeah, but what's the consequence? It doesn't seem to actually add much overhead. I wouldn't be surprised if Javascript automatically optimizes this (part of why JS can be used as a functional language). I did try to search for this, and [this person](https://stackoverflow.com/a/6476564/733862) seems to think there's no problem using inner functions, that "the convenience inner functions afford completely outweighs the small run time overhead" and they should only be moved out if there's a performance issue. Additionally, the next comment on that post shows a jsfiddle example with benchmarks.

Reply:
> Is there a reason to not use this syntax to deconstruct props into variables

As you can see, I do that for `shouldDeemphasize`, but since `repositoryHeading` needs to pass the props one function deeper to `shouldDeemphasize`, it's convenient to have the props in a variable.

Reply:
> How do we test those functions inside of render?

The functions inside render all affect what's rendered in the component. While we're not unit testing them, we are using [snapshots](https://jestjs.io/docs/en/snapshot-testing) to ensure the rendered component does not unexpectedly change (just added for this component ;). There's also an argument to be made that these nested functions are similar to private methods in Ruby. We don't test them because they are not public interfaces, only implementation details (not everything is a white box).

Reply:
> no real separation of concerns
> violation of the single responsibility principle

This seems like only one point to me. :) Since all the functions are specifically related to what is rendered how (i.e. which css classes to add, etc.), I'm not sure I agree with these statements. It seems like exactly what render should be doing. The one function with non-render specific logic (`yStream()`) is a helper function outside the component.

> it should be concerned primarily with rendering things (per the name) and not how to render them (hiding functions inside)

Not sure I understand this distinction you're making. render is concerned with what to render (html) how (css classes).